### PR TITLE
FEATURE: Tagging responses

### LINKED ISSUE #396: [FEATURE] As a user I want to tag responses and filter by tag
**Is your feature request related to a problem? Please describe.**
When you receive a lot of submissions the response list can get messy and getting an overview is difficult.

**Describe the solution you'd like**
Allow tagging and filtering by tag.

**Note**
We built tagging for a previous iteration of Formbricks. However, since then a few things have changed:
- We switched Tailwind Headless UI with [shadcn interpretation of Radix.](https://ui.shadcn.com)
- We added a few more data layers.
- Tags should be connected with **Products** so that teams can use the same tags on prod and dev environments.

This is the [Tagging.tsx](https://github.com/formbricks/formbricks/blob/c13547af1adda99a622b1882480b18dfe72d5635/apps/web/src/components/forms/shared/Tagging.tsx), this is a [file where we used it](https://github.com/formbricks/formbricks/blob/c13547af1adda99a622b1882480b18dfe72d5635/apps/web/src/components/forms/pmf/PMFTimeline.tsx), here is how we [filtered by tags](https://github.com/formbricks/formbricks/blob/c13547af1adda99a622b1882480b18dfe72d5635/apps/web/src/components/forms/shared/FilterNavigation.tsx) and this is [the overall commit](https://github.com/formbricks/formbricks/tree/c13547af1adda99a622b1882480b18dfe72d5635) in case you'd like to spin it up and see how it worked.

**Tasks**
1. Review the tagging and filter by tag functionality linked above
2. Previously, tags were an array of strings in the Response model. Decide if its better to add a model Tag to be able to edit and merge tags (see below)
3. Decide to use the [shadcn combobox ](https://ui.shadcn.com/docs/components/combobox)or this [fancy box](https://craft.mxkaske.dev/post/fancy-box) to build a tagging UX similar to this:
grafik
4. You're free to make the look & feel match our UI. It's not super essential so no need to get lost in details :)
5. In terms of functionality, refer to [this demo](https://demo.delighted.com/dashboard)
- Allow typing for a tag and filter the selection in the dropdown based on the typed string. On ENTER, if a tag is selected, select it.
- If you hit enter and no tag is selected in the dropdown, create a new one
6. Add a "Tags" page to manage tags in the Settings under Team -> Tags with a list of tags which contains:
- The tags name with an editable input field which auto updates (we have this functionality in the SurveyEditor.tsx for the survey name). When the name change was saved, show a toast.
- A count how often this tag was assigned.
- A delete button. Add toasts on success and err.
- A merge button: You can merge two tags together. The merge UI allows users to select a second tag and then the tag that was selected is merged into the first tag where the user hit the "Merge" button. You could wrap the Fancy or Combobox in a Tooltip which opens on "Merge" click to select the second tag. Make sure to add toasts.
7. Add a little settings icon on the right hand side at the bottom of the response card which links to the Tags settings page.

### PR DESCRIPTION
**Pull request by pandeymangg**
## What does this PR do?

Fixes #396  (issue)

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Chore (refactoring code, technical debt, workflow improvements)
- [ ] Enhancement (small improvements)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change adds a new database migration
- [ ] This change requires a documentation update

## How should this be tested?

- Test A
- Test B

## Checklist

- [x] Read the [contributing guide](https://github.com/formbricks/formbricks/blob/main/CONTRIBUTING.md)
- [x] Self-reviewed my own code
- [x] Commented on my code in hard-to-understand bits
- [x] Ran `pnpm build`
- [x] Checked for warnings, there are none
- [x] Removed all `console.logs`
- [ ] Merged the latest changes from main onto my branch with `git pull origin main`
- [ ] My changes don't cause any responsiveness issues
- [ ] Updated the Formbricks Docs if changes were necessary

### CODE REVIEW

**Comment by jobenjada** [comment]
Hey Pandeyman!

Thanks so much, we really appreciate the work you put into this 😊 

I'll provide feedback on the UI / UX side of things, Matti will review the code. Here is my Feedback:

1. I broke it quite easily: When you add a tag with a capital letter, it gets saved as lower case. This breaks the tagging.
grafik
https://tella.video/johanness-video-948o

2. When there are no tags, the UI is confusing: "No tags found" -> "Search tags" is contradictory
grafik
- Remove the icon
- Change placeholder tag: If no tags are found -> "Add tag..."
- If tags are found keep "Search tags..." but write "tag" with lower case

3. No hover indicator. Add `hover:bg-slate-50 hover:cursor-pointer` in the tag selection list:
grafik

4. Same here:
grafik

5. No accessibility: Impossible to navigate with keyboard. When there is no tag in the selection the new tag should be added on "Enter"
grafik

6. When you click out of the Search bar but leave text inside, the next time you hit "Add Tag" it looks like this:
grafik

7. Remove the Plus Icon, the layout itself already makes clear that this is a tag:
grafik


**Tag Page:**
grafik
1. This placeholder doesnt make sense here. It would be easier to display the empty table with a "Tag a submission to find your list of tags here."
2. The line isn't needed

grafik
3. Remove "tags". It suggests that there are several tags but the count is actually counting tagged responses. Just leave the number, should be clear.

4. Replace the plus icon with the Hashtag icon (heroicons)
grafik

5. Input field: Increase height a bit and make it fill the width of the cell. Always show the border (not only on hover) to make sure everyone understands that the tags can be changed there:
grafik

6. Increase spacing-x between actions. I know I said it differently before but make Delete of variant="alert" (or whatever the red button is :) and Merge variant "secondary

7. Remove not needed space:
grafik

Thanks again, big contribution! 🙌

**Review by mattinannt — CHANGES_REQUESTED** [review]
@pandeymangg Thanks for the feature PR 💪 Looks great 😊
I marked a few things in the code that I came across.

Additionally a few things that can be improved 😊:

**Product -> Environment**
After talking with Johannes and seeing the endpoints, I think it would make more sense to move the tags as children to the environments instead of the product. A user who creates or deletes a tag in the dev environment doesn't expect those changes to be reflected in the production environment and also most of the functionality as well as the current endpoint structure is built for everything to be a child of the environment.
Sorry for the extra work you have due to this decision 😥

**PR Description**
Also in the PR the `How to test` field is quite important to me because it makes it much easier to test the PR and find & test the functionality that has been changed, so please fill this in for the next PR 😊.

**Migrations**.
You also have a few migrations due to your journey of testing different approaches. It would be great if you could merge them all into one. What I usually do to achieve this:
```bash
git checkout main
git pull
# Reset the database to the current state of main
pnpm prisma migrate dev
git checkout [your-branch]
# delete all migrations you have created in the branch
# Then pull the latest changes from main
git pull origin main
# Create a new migration that is the difference between the database state of main and your changes to the schema
pnpm prisma migrate dev
```

**Prettier**
We use prettier to format the code to have consistent formatting in all files. Please install prettier in VSCode ( https://marketplace.visualstudio.com/items?itemName=esbenp.prettier-vscode ) and make sure it is your default formatter and that you have `format on save` enabled.
Also, the `pnpm format` command formats all code files according to the prettier project rules, and if your code is already formatted when you write it, this command shouldn't change anything.

File: apps/web/app/environments/[environmentId]/surveys/[surveyId]/responses/ResponseTimeline.tsx [hunk 1]
```diff
@@ -13,10 +13,12 @@ import { ArrowDownTrayIcon } from "@heroicons/react/24/outline";
 import { useState } from "react";
 import { toast } from "react-hot-toast";
 import { getTodaysDateFormatted } from "@formbricks/lib/time";
+import { useProduct } from "@/lib/products/products";
 
 export default function ResponseTimeline({ environmentId, surveyId }) {
   const { responsesData, isLoadingResponses, isErrorResponses } = useResponses(environmentId, surveyId);
   const { survey, isLoadingSurvey, isErrorSurvey } = useSurvey(environmentId, surveyId);
+  const { product } = useProduct(environmentId);
```
[the remark this answers is not in the document]
Comment (pandeymangg) [reply apps/web/app/environments/[environmentId]/surveys/[surveyId]/responses/ResponseTimeline.tsx:21]: Got it, I was just using it to get the product Id so I'll do that in the single response component

File: apps/web/pages/api/v1/environments/[environmentId]/product/[productId]/tags/[tagId]/index.ts [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+    try {
+      tag = await prisma.tag.delete({
+        where: {
+          id: tagId,
+        },
+      });
+    } catch (e) {
+      if (e instanceof Prisma.PrismaClientKnownRequestError) {
+        if (e.code === "P2002") {
+          return res.status(400).json({ message: "Tag already exists" });
```
[the remark this answers is not in the document]
Comment (pandeymangg) [reply apps/web/pages/api/v1/environments/[environmentId]/product/[productId]/tags/[tagId]/index.ts:92]: Got it, I'll make these changes and update the PR

**Review by mattinannt — APPROVED** [review]
@pandeymangg Thanks a lot for the changes 🤗 Works much better now 😊

Unfortunately, there were some build bugs on my end that I had to fix.

The behaviour of the tag combobox needed some improvement, as the user didn't have the option to create tags that are partials of existing tags (e.g. "Feature" could not be created if a tag "Feature Request" already exists). Also added case-insensitive search of tags.

I will merge the changes now 😊

**Pull request merged**